src/extrinsic/contracts.rs

```diff
@@ -17,122 +17,154 @@
 use codec::{Compact, Encode};
-use serde::de::DeserializeOwned;
-use sp_runtime::traits::GetRuntimeBlockType;
 
 pub const CONTRACTS_MODULE: &str = "Contracts";
-pub const CONTRACTS_PUT_CODE: &str = "put_code";
-pub const CONTRACTS_INSTANTIATE: &str = "instantiate";
-pub const CONTRACTS_INSTANTIATE_WITH_CODE: &str = "instantiate_with_code";
-pub const CONTRACTS_CALL: &str = "call";
-
-type Gas = u64;
-type Data = Vec<u8>;
-type Code = Vec<u8>;
-type Salt = Vec<u8>;
-
-type GasLimit = Compact<Gas>;
-type Endowment<Currency> = Compact<Currency>;
-type Value<Currency> = Compact<Currency>;
-
-pub type ContractPutCodeFn = (CallIndex, GasLimit, Data);
-pub type ContractInstantiateFn<Currency, Hash> =
-	(CallIndex, Endowment<Currency>, GasLimit, Hash, Data);
-pub type ContractInstantiateWithCodeFn<Currency> =
-	(CallIndex, Endowment<Currency>, GasLimit, Code, Data, Salt);
-pub type ContractCallFn<Address, Currency> = (CallIndex, Address, Value<Currency>, GasLimit, Data);
-
-pub type ContractPutCodeXt<Address, Signature, SignedExtra> =
-	UncheckedExtrinsicV4<Address, ContractPutCodeFn, Signature, SignedExtra>;
-pub type ContractInstantiateXt<Address, Signature, SignedExtra, Currency, Hash> =
-	UncheckedExtrinsicV4<Address, ContractInstantiateFn<Currency, Hash>, Signature, SignedExtra>;
-pub type ContractInstantiateWithCodeXt<Address, Signature, SignedExtra, Currency> =
-	UncheckedExtrinsicV4<Address, ContractInstantiateWithCodeFn<Currency>, Signature, SignedExtra>;
-pub type ContractCallXt<Address, Signature, SignedExtra, Currency> =
-	UncheckedExtrinsicV4<Address, ContractCallFn<Address, Currency>, Signature, SignedExtra>;
+pub const PUT_CODE: &str = "put_code";
+pub const INSTANTIATE: &str = "instantiate";
+pub const INSTANTIATE_WITH_CODE: &str = "instantiate_with_code";
+pub const CALL: &str = "call";
+
+pub type GasLimitFor<M> = Compact<<M as CreateContractsExtrinsic>::Gas>;
+pub type ValueFor<M> = Compact<<M as CreateContractsExtrinsic>::Currency>;
+pub type EndowmentFor<M> = Compact<<M as CreateContractsExtrinsic>::Currency>;
+pub type DataFor<M> = <M as CreateContractsExtrinsic>::Data;
+pub type CodeFor<M> = <M as CreateContractsExtrinsic>::Code;
+pub type SaltFor<M> = <M as CreateContractsExtrinsic>::Salt;
+pub type HashFor<M> = <M as CreateContractsExtrinsic>::Hash;
+pub type AddressFor<M> = <M as CreateContractsExtrinsic>::Address;
+
+/// Call for putting code in a contract.
+pub type PutCodeFor<M> = (CallIndex, GasLimitFor<M>, DataFor<M>);
+
+/// Call for instantiating a contract with the code hash.
+pub type InstantiateWithHashFor<M> =
+	(CallIndex, EndowmentFor<M>, GasLimitFor<M>, HashFor<M>, DataFor<M>);
+
+/// Call for instantiating a contract with code and salt.
+pub type InstantiateWithCodeFor<M> =
+	(CallIndex, EndowmentFor<M>, GasLimitFor<M>, CodeFor<M>, DataFor<M>, SaltFor<M>);
+
+/// Call for calling a function inside a contract.
+pub type ContractCallFor<M> = (CallIndex, AddressFor<M>, ValueFor<M>, GasLimitFor<M>, DataFor<M>);
```

Comment:
I'm wondering if we couldn't also "move" these types into the trait?
Do you think it's overkill or is there a specific reason not to do it?

Reply:
Sure, that's possible as well. If it helps to understand the code, I believe that's worth it. But maybe we should do that in a different PR, as this PR already introduces a lot of changes. What do you think?

Reply:
Works for me 👍

Reply:
Here you go: https://github.com/scs/substrate-api-client/issues/441 :)